### ISSUE TITLE
Unused pub fn in a private module triggers warning 

### ISSUE THREAD

**Issue by kpp** [rust-lang/rust#70104]
I tried this code:

```rust
pub(crate) mod a {
    pub fn f() { }
}

```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=6047c165fa563d155cfed7300d366387

I expected to see this happen: no warnings

Instead, this happened:

```bash
warning: function is never used: `f`
 --> src/main.rs:2:12
  |
2 |     pub fn f() { }
  |            ^
  |
  = note: `#[warn(dead_code)]` on by default
```

### Meta

`rustc --version --verbose`:
```
$ rustc -vV
rustc 1.42.0 (b8cedc004 2020-03-09)
binary: rustc
commit-hash: b8cedc00407a4c56a3bda1ed605c6fc166655447
commit-date: 2020-03-09
host: x86_64-unknown-linux-gnu
release: 1.42.0
LLVM version: 9.0

```

**Comment by Centril** [comment]
This seems like correct behavior to me; the function could never be reached outside the crate, and it's not used inside the crate, so it is dead code.

**Comment by kpp** [comment]
However it is reachable inside the crate and is used in `#[cfg(test)]` crate. Without `pub` it is unreachable, with `pub` it triggers warning, with `#[cfg(test)]` it looks ugly`.

**Comment by Centril** [comment]
Being reachable inside the crate is not a sufficient condition. It must also be used inside the crate by something that is not itself dead code, and here it is not.

**Comment by kpp** [comment]
So is it a well documented and known behaviour?

**Comment by Centril** [comment]
Well it's know and expected behavior; well documented idk...

**Comment by kpp** [comment]
Thanks!
